### PR TITLE
Detail Screen

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets
- [#Task 75921: Detail Screen](https://edu-redmine.sun-asterisk.vn/issues/xxxx)
## WHAT
- Contrast,Draw, filter, crop, and Download image
## Evidence (Screenshot or Video)

https://github.com/awesome-academy/HN_Android_2024_Image_Editor_2/assets/163242235/0581ba6b-6ac7-4c3d-94e5-e4ed013a7244

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/blob/master/eng/android/coding_convention_android_kotlin.md | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(Other notes)*

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: app/src/main/java/com/example/imageEditor2/custom/CropOverlayView.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    fun removeCropCallback() {
+        mCropSuccessCallback = null
+        paint.color = Color.TRANSPARENT
+        cropRect = RectF(0f, 0f, 0f, 0f)
+        invalidate()
+    }
+
+    fun haveInstanceListener() = mCropSuccessCallback != null
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/custom/CropOverlayView.kt]:
```suggestion
    fun hasInstanceListener() = mCropSuccessCallback != null
```

File: app/src/main/java/com/example/imageEditor2/custom/ImageListener.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
+package com.example.imageEditor2.custom
+
+import android.view.GestureDetector
+import android.view.MotionEvent
+import android.widget.ImageView
+import kotlin.math.abs
+
+class ImageListener :
+    GestureDetector.SimpleOnGestureListener() {
+    private var imageView: ImageView? = null
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/custom/ImageListener.kt]:
```suggestion
    private var mImageView: ImageView? = null
```

File: app/src/main/java/com/example/imageEditor2/custom/ImageListener.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        this.imageView = imageView
+    }
+
+    fun removeListener() {
+        imageView = null
+    }
+
+    fun haveListener() = imageView != null
+
+    private var isZoomIn = false
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/custom/ImageListener.kt:11]: Move lên top

File: app/src/main/java/com/example/imageEditor2/ui/detail/DetailViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.launch
+
+class DetailViewModel(private val detailRepository: DetailRepository) : BaseViewModel() {
+    private val _message = MutableLiveData<Int>()
+    val message: LiveData<Int> = _message
+
+    private val _error = MutableLiveData<String>()
+    val error: LiveData<String> = _error
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/ui/detail/DetailViewModel.kt]:
```suggestion
    private val mMessage = MutableLiveData<Int>()
    val message: LiveData<Int> = _message

    private val mError = MutableLiveData<String>()
    val error: LiveData<String> = _error
```

File: app/src/main/java/com/example/imageEditor2/ui/detail/ImageDetailActivity.kt [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+import org.koin.androidx.viewmodel.ext.android.viewModel
+
+class ImageDetailActivity :
+    BaseActivity<ActivityDetailImageBinding>(),
+    CropSuccessCallback,
+    OnFilterPicked {
+    private val mUrl by lazy { intent.getStringExtra(URL) }
+    private val mImageListener by lazy { ImageListener() }
+    private var mScaleGestureDetector: GestureDetector? = null
+    private var isFiltering = false
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/ui/detail/ImageDetailActivity.kt]: .

File: app/src/main/java/com/example/imageEditor2/ui/detail/ImageDetailActivity.kt [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+            strokeWidth = 5f
+            style = Paint.Style.STROKE
+            isAntiAlias = true
+        }
+
+    private var mLastTouchX = 0f
+    private var mLastTouchY: Float = 0f
+    private var croppedBitmap: Bitmap? = null
+    private val mListEmoji = mutableListOf<ImageView>()
+    private val viewModel: DetailViewModel by viewModel()
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/ui/detail/ImageDetailActivity.kt]: .

File: app/src/main/java/com/example/imageEditor2/ui/detail/ImageDetailActivity.kt [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+    private lateinit var emoji: String
+    private lateinit var filterAdapter: FilterAdapter
```
Comment (daolq-2712) [inline app/src/main/java/com/example/imageEditor2/ui/detail/ImageDetailActivity.kt]: .

**Review by daolq-2712 — CHANGES_REQUESTED** [review]

**Review by daolq-2712 — APPROVED** [review]
OK!

**Pull request merged**